Brainy Quote: Shows more than one quote if requested.

MyGitUsername opened the pull request:
## Description

If the user requests 'quotes' or 'quotations' 3 results are returned and displayed in a list view.  The brainy quote api is called 3 seperate times in order to obtain 3 quotes.  In may be prudent to check if any of the quotes are the same and that there are 3 unique quotes for the search query.

## Related Issues and Discussions
Fixes #3279.  

## People to notify
@pjhampton 

Instant Answer Page: https://duck.co/ia/view/brainy_quote

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

pjhampton commented:
This looks pretty good so far. I need to do some more testing before we can push this on 👍 

![screen shot 2017-05-16 at 9 42 53 pm](https://cloud.githubusercontent.com/assets/8960296/26127666/e1a5524a-3a81-11e7-84fb-cf60732dcbce.png)
![screen shot 2017-05-16 at 9 43 10 pm](https://cloud.githubusercontent.com/assets/8960296/26127667/e1afe534-3a81-11e7-9b17-0e2d4ed70ddd.png)
![screen shot 2017-05-16 at 9 42 59 pm](https://cloud.githubusercontent.com/assets/8960296/26127668/e1b3ca6e-3a81-11e7-8e12-7f8854e847a8.png)

MyGitUsername commented:
Great - thank you for reviewing.

gautamkrishnar commented:
Looking cool :heart:

ManrajGrover commented on share/spice/brainy_quote/brainy_quote.js:
```diff
@@ -1,32 +1,89 @@
-                    moreAt: true
+	
+	if (!DDG.get_query().match(/quotes|quotations/)) {
+            Spice.add({
```
We can move common parts from Spice object here and on [line 66](https://github.com/duckduckgo/zeroclickinfo-spice/pull/3288/files#diff-837555897925643aabbcf8c8d5071a1fR66) to a separate object above the plural check and only have customisations in `if` blocks before adding to Spice.

MyGitUsername replied:
Thank you for the suggestion.  I made this change and submitted a commit.

sahildua2305 commented:
@pjhampton feel free to merge this once you're happy with this completely 😄

moollaza commented on share/spice/brainy_quote/brainy_quote.js:
```diff
@@ -1,32 +1,86 @@
+                })
+                .fail(function( jqxhr, textStatus, error ) {
+                    var err = textStatus + ", " + error;
+                    console.log( "Request Failed: " + err );
```
Please remove. We don't log errors to the console. This should silently fail.

moollaza commented on share/spice/brainy_quote/brainy_quote.js:
```diff
@@ -1,32 +1,86 @@
-                    moreAt: true
+        };
+	
+	if (!DDG.get_query().match(/quotes|quotations/)) {
```
nit: Please use 4-space soft tabs, we don't use hard tabs

moollaza commented on share/spice/brainy_quote/brainy_quote.js:
```diff
@@ -1,32 +1,86 @@
             }
-        });
+
+            jqxhr.done(function() {
```
The above loop overwrites the jqxhr object, so I believe this is only going to wait for the last request to finish? AFAIK you need to use `$.when` to wait for multiple deferred objects to return. I don't think this is correct...

https://api.jquery.com/jQuery.when/

MyGitUsername replied:
Thank you for the suggestions.  Happy to contribute and learn from you and the rest of the DDH team.

moollaza commented:
@MyGitUsername let's also add some CSS to ensure the first and last list items don't have top/bottom borders. There should just be a border between items 👍

MyGitUsername commented on lib/DDG/Spice/BrainyQuote.pm:
```diff
@@ -9,10 +9,16 @@ triggers startend => 'quote', 'quotes', 'quotation', 'quotations';
 spice to => 'https://www.brainyquote.com/api/ddg?q=$1';
 spice wrap_jsonp_callback => 1;
 
+spice alt_to => {
```
Can you please clarify further?  When I call $.getJSON("/ddg/spice/brainy_quote/" +  query) I receive an object with HTML response text.

moollaza replied:
Ah, my mistake -- The `alt_to` is necessary to force a JSON response, instead of JSONP.

The name should be something better than `test_endpoint` though 👍

moollaza commented on share/spice/brainy_quote/brainy_quote.js:
```diff
@@ -1,32 +1,76 @@
+            if (arg2 === undefined) {
+                return;
+            }
+            return arg1 === undefined ? arg2.replace(/ quote$/, "") : arg1;
```
this could actually be further simplified: 

`return arg1 || arg2.replace(/ quote$/, "")`

MyGitUsername commented:
@moollaza In addition to addressing your comments, the latest push includes additional CSS code that makes the list view result look more like the base result.  Specifically, each quote is preceded by quotation marks and the margin between the first quote and the window-length border is reduced.  Please let me know if you have any other suggestions.

moollaza commented:
> @moollaza the latest push includes additional CSS code that makes the list view result look more like the base result. Specifically, each quote is preceded by quotation marks and the margin between the first quote and the window-length border is reduced.

Sorry for the delay, I've been away on vacation! The changes are deployed to beta but they don't seem to be working well. I would actually prefer to reduce the amount of custom CSS and UI here to simplify this IA as much as possible. I would opt for removing all the additional styling CSS so this IA looks like others that use the list view -- the custom UI has always been problematic especially on mobile.

moollaza commented:
@MyGitUsername I had some time to kill so I tried to see if I could get this IA to look/work a little more like I expected.

I didn't get it finished, but it's nearly working. I've attempted to reduce the visual complexity, add a title, and determine when we have multiple quotes from the same author and set the title accordingly, rather than re-iterating for same author for each quote.

My code is here: https://gist.github.com/moollaza/252eeaa3315b312e6cc4bf58c4237088

Would you like to pickup where I've left off and finish up this IA?

MyGitUsername commented:
@moollaza I am happy to continue working on this IA.   What is the best way to get your changes so I can pickup where you left off?

moollaza commented:
@MyGitUsername you should be able to checkout my branch and merge it into your, or simply cherry-pick that last commit. My branch is a fork of yours that I added a commit to 👍  -- Does that make sense?

MyGitUsername commented:
@moollaza that does make sense.  Yet, I am a bit rusty with git and am having difficulties trying to cherry-pick your commit and checkout the branch pr/3288.  I am available on slack, if you have time to guide me through the git workflow.  Thanks.

kirkins commented:
@MyGitUsername Try this and let me know if it works:

```
git remote add moollaza https://github.com/moollaza/zeroclickinfo-goodies.git
git fetch moollaza
git checkout brainy-quote-improvement moollaza/pr/3288
```

MyGitUsername commented:
@kirkins Thanks for quick response.  No luck.  

Here is a screenshot of the 'failed' console output:

![screenshot from 2017-06-09 07 35 40](https://user-images.githubusercontent.com/7421657/26980288-88d47ad4-4ce6-11e7-9df2-6eeb71f59a04.png)

kirkins commented:
@MyGitUsername Sorry, I edited my previous comment. Let me know if that works for you.

kirkins commented:
@MyGitUsername also I think I made the mistake of thinking your branches name is `mbrainy-quote-improvement` when it is actually `mygitusername/brainy-quote-improvement`.

So use `mygitusername/brainy-quote-improvement` instead if that's the name of your branch.

kirkins commented:
@MyGitUsername when I look at your fork I don't see the branch, did you delete it?

If so use this to clone Moollaza's as a new one with the same name _(difference is the -b flag to create new branch)_:

`git checkout -b mygitusername/brainy-quote-improvement moollaza/pr/3288`

MyGitUsername commented:
@kirkins `git checkout -b pr/3288 moollaza/pr/3288` is failing with the following error:
```
fatal: Cannot update paths and switch to branch 'pr/3288' at the same time.
Did you intend to checkout 'moollaza/pr/3288' which can not be resolved as commit?
```
Looking at the console output of `fetch moollaza`, it seems to me that this command is failing to fetch the pr/3288 branch.

MyGitUsername commented:
@kirkins I haven't 'knowingly' deleted my branch mygitusername/brainy-quote-improvement.

In fact, running `git status` returns:

`On branch mygitusername/brainy-quote-improvement
nothing to commit, working tree clean`

MyGitUsername commented:
@kirkins Also, your suggested workflow states `git remote add moollaza https://github.com/moollaza/zeroclickinfo-goodies.git`

I would suspect that pr/3288 branch would be in the spice repository, so accordingly the command should be `git remote add moollaza https://github.com/moollaza/zeroclickinfo-goodies.git`.  

However, running `fetch moollaza` after the latter command is also not fetching the pr/3288 branch.

kirkins commented:
@MyGitUsername you're right I meant to add spice as the remote. And now that I checkout your branch I can see you didn't delete it.

I'm trying it myself and having issues too. I'm trying to list all moollaza's branches `git ls-remote --heads moollaza` but I don't see it. Must be a better way...

kirkins commented:
@MyGitUsername sorry for so many messages but I'm also noticing that the commit that @moollaza linked to isn't on his branch. It's on the duckduckgo branch https://github.com/duckduckgo/zeroclickinfo-spice/commit/853f7a8a2d61a7b8e9c0118d4bca35a13120da56

But I still haven't been able to figure out how to pull it into your repo. I'm actually really curious to figure out how to do this now :s

MyGitUsername commented:
@kirkins Because the branch is store on the duckduckgo repo try:

`git fetch upstream` 
then you can checkout out the branch or cherry-pick a commit from it
'git cherry-pick 853f7a8'

That worked for me.

MyGitUsername commented:
@moollaza I finished up where you left off and fixed the title.  Reviewing your recent comment with a list of fixes it seems to me that the visual complexity has be reduced, and a title has been added and set accordingly.

Please let me know if there were any other fixes you would like to make.

moollaza commented:
@MyGitUsername Thanks! Excellent work! 💯  This LGTM :+1:
